internal/uvm/scsi/README.md

```diff
@@ -0,0 +1,101 @@
+- Tracks what SCSI devices are currently attached, along with what controllers/LUNs are
+  used.
+- When it is asked to attach a new SCSI device, it will first check if the attachment
+  already exists, and increase its refcount if so. If not, it will allocate a new
+  controller/LUN slot for the attachment, and then call the `Attacher` to actually carry
+  out the attach operation.
+- When it is asked to detach a SCSI device, it first uses the `Unplugger` to carry out any
+  guest-side remove actions, and then uses the `Attacher` to remove the attachment from
+  the VM.
+- Tracks refcount on any attached SCSI devices, so that they are not detached until there
+  has been a detach request for each matching attach request.
+
+`mountManager`:
+- Tracks current SCSI devices mounted in the guest OS, and what mount options were applied.
+- When it is asked to mount a new SCSI device, it will first check if the mount (with same options)
+  already exists, and increase its refcount if so. If not, it will track the new mount, and
+  call the `Mounter` to actually carry out the guest mount operation.
+- Tracks refcount on any mounted SCSI devices, so that they are not unmounted until there has
+  been an unmount request for each matching mount request.
+
+### Low level, the `Attacher`, `Mounter` and `Unplugger` types
+
+These three are interfaces defined by the package. They provide a way for the client to
+provide low-level implementations of SCSI operations, passed in when constructing the
+`Manager` object.
+
+These types are what carry out the actual operations on HCS or the bridge, for instance, to do
+a SCSI attachment or mount.
+
+The `Unplugger` type exists to help with removing a SCSI device from the guest before detaching. It
+is used in the case of Linux containers to cleanly unplug the SCSI device.
+
+## Future work
+
+Some thoughts on how this package could evolve in the future. This is intended to inform the direction
+of future changes as we work in the package.
+
+- The `mountManager` actually has very little to do with SCSI (at least for Linux containers, Windows
+  may be more complicated/limited). In fact, the only SCSI-specific part of mounting in the guest is
+  pretty much just identifying a block device based on the SCSI controller/LUN. It would be interesting
+  to separate out the SCSI controller/LUN -> block device mapping part from the rest of the guest mount
+  operation. This could enable us to e.g. use the same "mount" operation for SCSI and VPMEM, since they
+  both present a block device.
+- We should not be silently and automatically scanning a SCSI device for verity info. Determining what
+  (if any) verity info to use for a device should probably be determined by the client of the package.
+- Likewise, ACLing disks so a VM can access them should likely fall outside the purview of the package
+  as well.
+- The various implementations of `Attacher`, `Mounter`, and `Unplugger` should probably live outisde
```

Comment:
I'm not hugely opinionated on this. I don't think ACLing needs to fall specifically under the `scsi` package. However, there is room for a higher level `StorageManager` or similar type which handles "storage" (attaches/mounts of SCSI/VPMem/etc.) to the UVM. Perhaps `scsi.Manager` can just grow into this over time, in which case it's reasonable to keep ACLing in there.

In case you missed it, I did have ACLing part of `scsi.Manager` originally. The current design has some experimental changes that I am looking for feedback on. See comment [here](https://github.com/microsoft/hcsshim/pull/1744#discussion_r1178501433) for details.

Reply:
Also, for context, the previous scsi implementation did not have any check (AFAICT) to only add an ACL if the disk wasn't already attached. It appears to add the ACL every time `AddSCSI` is called.